File: small_problems/medium1/word_to_digit.rb
# Word to Digit
# Write a method that takes a string as input and returns the same string 
# with any sequence of the words zero, one, two, ..., nine converted to a 
# string of digits.

# Input: A String containing a sentence (space-separated words)
# Output: The same string, but any instances of number words (eg. 'one')
#   are replaced with the corresponding string digit.

# Example:
# str = 'Please call me at five five five one two three four. Thanks.'
# word_to_digit(str) == 'Please call me at 5 5 5 1 2 3 4. Thanks.'

# Data: 
#   The string given as input 
#   - Maybe split string into words?
#   Maybe use gsub!
#   - gsub! requires a pattern (regex)
#   /\b#{var}\b/ - Matches `var` (containing some word) in a longer string
#   Probably need some data structure to hold the word-digit conversions
# 



# Algorithm:
# Given a string as input, sentence:
# Create a hash DIGITS to store the word-digit conversions
#   eg. { 'one' => '1' }

# Go through sentence one word at a time. For each word:
#   If the current word is one of the keys in DIGITS, replace it
#     with the corresponding value from DIGITS[word] 
#     eg. 'five' is a key in DIGITS; replace 'five' with DIGITS['five'] => '5'
#   (Modify sentence in place)

# Digits Approach:
#   Iterate through DIGITS; For each key-value pair:
#     Run through sentence and replace all instances of the key with the value
#   Return sentence

# Without gsub!:
# Split sentence into words. Iterate through words; for each word:
#   Clean the current word (non-destructive) 
#   - Remove the last character if it is non-alphabetic.
#   If the cleaned word is a key in DIGITS, replace the current word
#     with the corresponding digit 
# * You can give String#[]= a substring, and it will replace the first matching
#   substring with the new string.

#     - only the part of the word that matches the key.
#     eg. 'four.' => 'four' (clean; original still 'four.')
#         'four.'['four'] => '4' => '4.'
#  => Array of words, numbers -> digits
#   Join array of words to a string (new string tho)

DIGITS = ['zero', 'one', 'two', 'three', 'four', 'five',
  'six', 'seven', 'eight', 'nine'].zip(Array('0'..'9')).to_h

ALPHABET = Array('a'..'z') + Array('A'..'Z')

# Using gsub! and regex
def word_to_digit(str)
  DIGITS.each do |word, digit|
    str.gsub!(/\b#{word}\b/, digit)
  end
  str
end


def word_to_digit(str)
  str.split.each do |word|
    cleaned_word = clean_word(word)
    if DIGITS.key?(cleaned_word.downcase)
      word[cleaned_word] = DIGITS[cleaned_word.downcase]
    end
  end.join(' ')
end

def clean_word(word)
  chars = word.chars
  chars.pop until ALPHABET.include?(chars.last)
  chars.join
end

str = 'Please call me at five five five one two three four. Thanks.'
p word_to_digit(str) == 'Please call me at 5 5 5 1 2 3 4. Thanks.'

File: javaScriptExercises/medium/wordToDigit.js
// Word to Digit
// - Write a function that takes a string (sentence) as argument and returns the
//   same string with every occurrence of a 'number' word ('zero', 'one, 'two',
//   etc.) converted to its corresponding digit number.

// Questions:
// 1) Is it possible for number words to be decimals and/or negative?
//    eg. "negative one" or "one point five"
// => Not right now (further exploration)

// 2) Is there a maximum possible value for number words? 
//    eg. Do we need to handle "fifty five", "one hundred", etc.? Or only 0-9?
// => Only 0-9

// 3) What should be done about compound words (eg. four-score)?
// => Leave them as-is (do not replace)

// 4) Should case be insensitive? (eg. One => 1)
// => Yes

// Input: 
// - A String (sentence)
// Output:
// - A new String: same sentence, but with number words (zero-nine) replaced with
//   the corresponding digits (0-9)

// Data: 
// `sentence` - A string of words (space-separated)
// > Split sentence into words based on spaces
// * This might result in some words being attached to their trailing punctuation
//   eg. 'two three four.' => ['two', 'three', 'four.']

// Create a dictionary mapping numeric words to their digit forms:
// eg. { 'zero': 0, 'one': 1, 'two': 2, ... }

// Iterate over words. For each word:
// -- Transform the word; if it's a standalone number word, convert it to a 
//    digit; otherwise, leave it as-is

// 1) Standardize the casing of the word to lowercase
// 2) Extract the word from the punctuation
//    > Split on word boundaries
//    ** If the word is compound (eg. four-score), we do NOT want to replace the
//       numeric word.
// 3) Check if the extracted word exists in the number dictionary; if it does,
//    replace it with the dictionary[word] value + punctuation. Otherwise, replace
//    it with word + punctuation.
// 4) => Array of transformed words => Join and return

function wordToDigit(sentence) {
  const NUMBER_DICTIONARY = { 
    zero: 0, one: 1, two: 2, three: 3, four: 4,
    five: 5, six: 6, seven: 7, eight: 8, nine: 9,
  }

  let words = sentence.split(' ');

  return words.map(word => {
    let [ wordOnly, punctuation, ...other ] = word.toLowerCase().split(/\b/);
    
    if (NUMBER_DICTIONARY.hasOwnProperty(wordOnly) && other.length === 0) {
      punctuation ||= '';
      return NUMBER_DICTIONARY[wordOnly] + punctuation;
    } else {
      return word;
    }
  }).join(' ');
}

// Alternative Approach (Given): 
// - Instead of iterating through word, iterate through NUMBER DICTIONARY and 
//   replace all occurrences of each number word with the corresponding digit.

function wordToDigit(sentence) {
  const NUMBER_DICTIONARY = { 
    zero: 0, one: 1, two: 2, three: 3, four: 4,
    five: 5, six: 6, seven: 7, eight: 8, nine: 9,
  }

  Object.keys(NUMBER_DICTIONARY).forEach(numericWord => {
    let searchPattern = new RegExp(`(?<!\\w[^ ])\\b${numericWord}\\b(?![^ ]\\w)`, 'ig');
    sentence = sentence.replaceAll(searchPattern, NUMBER_DICTIONARY[numericWord]);
  });

  return sentence;
}

// Examples:
// - Simple Example
console.log(wordToDigit('one two three four')); // 1 2 3 4
// - Case-insensitive
console.log(wordToDigit('One two thRee FOUR')); // 1 2 3 4
// - Can handle punctuation without removing it
console.log(wordToDigit('One! two!! thRee!!! FOUR!!!!')); // 1! 2!! 3!!! 4!!!!

// Compound (hyphen-joined) words involving numbers are not replaced (only standalones)
console.log(wordToDigit('four-score and seven years ago'));
// => "four-score and 7 years ago"
console.log(wordToDigit('score-four and seven years ago'));
// => "score-four and 7 years ago"

// Number words that happen to be part of another word are not replaced
console.log(wordToDigit('weight eight')); // 'weight 8'

// - Given Example
console.log(wordToDigit('Please call me at five five five one two three four. Thanks.'));
// "Please call me at 5 5 5 1 2 3 4. Thanks."

